src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -2018,72 +2000,146 @@ static void compaction_with_stealing_work(TaskTerminator* terminator, uint worke
+  {
+    assert(start < end, "precondition");
+    assert(mark_bitmap()->find_obj_beg(start, end) == end, "precondition");
+    HeapWord* bottom = _space_info[old_space_id].space()->bottom();
+    if (start != bottom) {
+      HeapWord* obj_start = mark_bitmap()->find_obj_beg_reverse(bottom, start);
+      HeapWord* after_obj = obj_start + cast_to_oop(obj_start)->size();
+      assert(after_obj == start, "precondition");
     }
+  }
+#endif
 
-    // Once a thread has drained it's stack, it should try to steal regions from
-    // other threads.
-    compaction_with_stealing_work(&_terminator, worker_id);
+  CollectedHeap::fill_with_objects(start, pointer_delta(end, start));
+  HeapWord* addr = start;
+  do {
+    size_t size = cast_to_oop(addr)->size();
+    start_array(old_space_id)->update_for_block(addr, addr + size);
+    addr += size;
+  } while (addr < end);
+}
+
+void PSParallelCompact::fill_dead_objs_in_dense_prefix(uint worker_id, uint num_workers) {
+  ParMarkBitMap* bitmap = mark_bitmap();
+
+  HeapWord* const bottom = _space_info[old_space_id].space()->bottom();
+  HeapWord* const prefix_end = dense_prefix(old_space_id);
 
-    // At this point all regions have been compacted, so it's now safe
-    // to update the deferred objects that cross region boundaries.
-    cm->drain_deferred_objects();
+  if (bottom == prefix_end) {
+    return;
   }
-};
+
+  size_t bottom_region = _summary_data.addr_to_region_idx(bottom);
+  size_t prefix_end_region = _summary_data.addr_to_region_idx(prefix_end);
+
+  size_t start_region;
+  size_t end_region;
+  split_regions_for_worker(bottom_region, prefix_end_region,
+                           worker_id, num_workers,
+                           &start_region, &end_region);
+
+  if (start_region == end_region) {
+    return;
+  }
+
+  HeapWord* const start_addr = _summary_data.region_to_addr(start_region);
+  HeapWord* const end_addr = _summary_data.region_to_addr(end_region);
+
+  // Skip live partial obj (if any) from previous region.
+  HeapWord* cur_addr;
+  RegionData* start_region_ptr = _summary_data.region(start_region);
+  if (start_region_ptr->partial_obj_size() != 0) {
+    HeapWord* partial_obj_start = start_region_ptr->partial_obj_addr();
+    assert(bitmap->is_marked(partial_obj_start), "inv");
+    cur_addr = partial_obj_start + cast_to_oop(partial_obj_start)->size();
+  } else {
+    cur_addr = start_addr;
+  }
+
+  // end_addr is inclusive to handle regions starting with dead space.
+  while (cur_addr <= end_addr) {
+    // Use prefix_end to handle trailing obj in each worker region-chunk.
+    HeapWord* live_start = bitmap->find_obj_beg(cur_addr, prefix_end);
+    if (cur_addr != live_start) {
+      // Only worker 0 handles proceeding dead space.
+      if (cur_addr != start_addr || worker_id == 0) {
+        fill_range_in_dense_prefix(cur_addr, live_start);
+      }
+    }
+    if (live_start >= end_addr) {
+      break;
+    }
+    assert(bitmap->is_marked(live_start), "inv");
```

Comment:
Does this conditional break statement just let the assertion pass? If so, what about removing the conditional statement and changing the assertion to `bitmap->is_marked(live_start) || live_start == prefix_end`.

Reply:
> The early-return is needed to ensure `cast_to_oop(live_start)->size()` is safe.

Got  it. Thanks.